src/google/protobuf/descriptor.cc

```diff
@@ -5487,6 +5497,112 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  FileDescriptor::Syntax syntax = result->file()->syntax();
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers
+  // custom JSON names
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, false);
```

Comment:
What's the purpose of this check?  If custom json_names are set why do we still warn/error?  I would expect setting an explicit json_name to be a good mitigation to overlapping field names

Reply:
I left that in place due to this comment from @acozzette on an issue I filed:
https://github.com/protocolbuffers/protobuf/issues/7192#issuecomment-695024065

And, indeed, the [custom JSON format](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#json-encoding-of-field-masks) for `FieldMask` simply converts the contained field names to camel-case names, without knowledge of the fields' custom names (e.g. field masks always refer to fields' _default_ JSON name).

I would personally _love_ to remove the check, since I agree that a custom JSON name seems like the best way to prevent the ambiguity it is complaining about. 🤷

Reply:
Ah that makes sense, thanks for the explanation!  This might be a good thing to follow up on, since that error case is still fairly confused (it does say "default" vs "default" clashes, but it's not clear why that matters).  The specific case I hit was a single proto2 file that:

1) had two fields with clashing defaults
2) didn't care about JSON at all
3) was compiled in a way where warnings triggered errors

Would it make sense to just disable the warning in the non-custom branch?

Reply:
Sure, disabling the warning for that case seems fair.

Reply:
Done

Reply:
I think this may not be necessary.  I'm nearly green for your original change, which doesn't include the oneof exemption and warns in proto2